cmd/dra-example-kubeletplugin/state.go

```diff
@@ -173,3 +193,67 @@ func (s *DeviceState) prepareDevices(claim *resourceapi.ResourceClaim) (Prepared
 func (s *DeviceState) unprepareDevices(claimUID string, devices PreparedDevices) error {
 	return nil
 }
+
+func (pds PreparedDevices) GetDevices() []*drapbv1.Device {
+	var devices []*drapbv1.Device
+	for _, pd := range pds {
+		devices = append(devices, &pd.Device)
+	}
+	return devices
+}
+
+type PtrHasGroupVersionKind[T any] interface {
+	*T
+	GroupVersionKind() schema.GroupVersionKind
+}
+
+// GetMergedOpaqueDeviceConfig merges configs across device classes and resource claims for a given device request.
+func GetMergedOpaqueDeviceConfig[T any, TI PtrHasGroupVersionKind[T]](request string, claim *resourceapi.ResourceClaim, classes map[string]*resourceapi.DeviceClass) (T, error) {
+	// Collect all configs in reverse priority order. Device class configs have
+	// lower priority than resource claim configs. Within either a device class
+	// or a resource claim, configs earlier in the list have lower priority
+	// than configs later in the list.
+	var allConfigs []resourceapi.DeviceConfiguration
+	for _, config := range classes[request].Spec.Config {
+		allConfigs = append(allConfigs, config.DeviceConfiguration)
+	}
+	for _, config := range claim.Spec.Devices.Config {
+		if !slices.Contains(config.Requests, request) {
+			continue
+		}
+		allConfigs = append(allConfigs, config.DeviceConfiguration)
+	}
+
+	// Merge all configs from the allConfigs slice in the order they were
+	// added. Fields in a config that is unmarshaled earlier will be
+	// overwritten by fields in a config that is unmarhaled later.
+	var mergedConfig T
+	for _, config := range allConfigs {
+		if config.Opaque == nil {
+			continue
+		}
+		if config.Opaque.Driver != DriverName {
+			continue
+		}
+
+		if err := json.Unmarshal(config.Opaque.Parameters.Raw, &mergedConfig); err != nil {
+			return *new(T), fmt.Errorf("error converting config parameters to GpuConfig: %v", err)
+		}
+
+		gvk := TI(&mergedConfig).GroupVersionKind()
+
+		if gvk.Group != configapi.GroupName {
+			return *new(T), fmt.Errorf("wrong group for config parameters: %v", gvk.Group)
+		}
+
+		if gvk.Version != configapi.Version {
+			return *new(T), fmt.Errorf("wrong version for config parameters: %v", gvk.Version)
+		}
+
+		if gvk.Kind != configapi.GpuConfigKind {
+			return *new(T), fmt.Errorf("wrong kind for config parameters: %v", gvk.Kind)
+		}
```

Comment:
It's not clear under which conditions these errors would occur. Factoring out into a validate function may also make sense.

Reply:
The function signature has changed to rely on `runtime.Scheme` to do validation for us. Take a look again and let me know if you still have questions.